src/Utility/HttpResponsePaginator.php

```diff
@@ -0,0 +1,284 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Auth0\SDK\Utility;
+
+use Auth0\SDK\Exception\NetworkException;
+use Auth0\SDK\Exception\PaginatorException;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Class HttpResponsePaginator.
+ */
+final class HttpResponsePaginator implements \Countable, \Iterator
+{
+    private HttpClient $httpClient;
+
+    private int $position = 0;
+    private int $requestLimit = 0;
+    private int $requestTotal = 0;
+    private int $requestCount = 0;
+    private array $results = [];
+
+    public function __construct(
+        HttpClient $httpClient
+    ) {
+        $this->httpClient = $httpClient;
+        $this->processLastResponse();
+    }
+
+    public function countNetworkRequests(): int
+    {
+        return $this->requestCount;
+    }
+
+    /**
+     * Return the total number of results available, according to the API.
+     */
+    public function count(): int
+    {
+        return $this->requestTotal;
+    }
+
+    /**
+     * Return the current result at our position, if available.
+     */
+    public function current()
+    {
+        return $this->valid() ? $this->result() : false;
+    }
+
+    /**
+     * Retrieve the current position, if valid.
+     */
+    public function key(): int
+    {
+        return $this->valid() ? $this->position : null;
+    }
+
+    /**
+     * Increase our position cursor.
+     */
+    public function next(): void
+    {
+        ++$this->position;
+    }
+
+    /**
+     * Return true if a result is available. If a result is not immediately available (cached) but the current position is less than the API-reported total results, a paginated network request will be attempted to get the next results. Returns false when no results are available at the current position.
+     */
+    public function valid(): bool
+    {
+        // A cached result is available.
+        if ($this->result()) {
+            return true;
+        }
+
+        // No cached result available; is our position beyond the API's reported total?
+        if ($this->position < $this->requestTotal) {
+            // No, there should be more results available for request. Do that.
+            return $this->getNextResults();
+        }
+
+        return false;
+    }
+
+    /**
+     * Reset position to 0.
+     */
+    public function rewind(): void
+    {
+        $this->position = 0;
+    }
+
+    /**
+     * Return the current result at our position, if available.
+     */
+    private function result()
+    {
+        return $this->results[$this->position] ?? null;
+    }
+
+    /**
+     * Set an HttpRequest's pagination params to safe defaults. Triggered when a HttpResponse is provided that isn't paginated, or doesn't have include_totals set (required for iteration.)
```

Comment:
What happens when you test this against `/logs` which only does checkpoint pagination?

https://auth0.com/docs/logs/retrieve-log-events-using-mgmt-api